fail2ban/server/banmanager.py

```diff
@@ -319,22 +311,39 @@ def _inBanList(self, ticket):
 	# @return the list of ticket to unban
 	
 	def unBanList(self, time):
-		try:
-			self.__lock.acquire()
+		with self.__lock:
 			# Permanent banning
 			if self.__banTime < 0:
 				return list()
 
-			# Gets the list of ticket to remove.
-			unBanList = [ticket for ticket in self.__banList if ticket.isTimedOut(time, self.__banTime)]
-			
+			# Check next unban time:
+			if self.__nextUnbanTime > time:
+				return list()
+
+			# Gets the list of ticket to remove (thereby correct next unban time).
+			unBanList = {}
+			self.__nextUnbanTime = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFL
+			for fid,ticket in self.__banList.iteritems():
+				# current time greater as end of ban - timed out:
+				eob = ticket.getEndOfBanTime(self.__banTime)
+				if time > eob:
+					unBanList[fid] = ticket
+				elif self.__nextUnbanTime > eob:
+					self.__nextUnbanTime = eob
+
 			# Removes tickets.
-			self.__banList = [ticket for ticket in self.__banList
-							  if ticket not in unBanList]
+			if len(unBanList):
+				if len(unBanList) / 2.0 <= len(self.__banList) / 3.0:
+					# few as 2/3 should be removed - remove particular items:
+					for fid in unBanList.iterkeys():
+						del self.__banList[fid]
```

Comment:
I know, several things were partially cherry-picked from my new ban-time-incr, where it all test-covered. But the tests are not completely compatible, will try to porting it.